File: 2026/09-September/07-Monday/solution.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        MOD = 10**9 + 7
        dp = 0
        last = {}

        for ch in s:
            new_dp = (2 * dp + 1) % MOD

            if ch in last:
                new_dp = (new_dp - last[ch]) % MOD

            last[ch] = (dp + 1) % MOD
            dp = new_dp

        return dp

File: 2026/09-September/07-Monday/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int distinctSubseqII(string s) {
        const int MOD = 1e9 + 7;

        long long dp = 0;
        vector<long long> last(26, 0);

        for (char ch : s) {
            long long newDp = (2 * dp + 1) % MOD;

            int idx = ch - 'a';
            newDp = (newDp - last[idx] + MOD) % MOD;

            last[idx] = (dp + 1) % MOD;
            dp = newDp;
        }

        return dp;
    }
};
